mars/services/task/execution/ray/executor.py

```diff
@@ -527,3 +529,86 @@ def _get_subtask_output_keys(chunk_graph: ChunkGraph):
             else:
                 output_keys[chunk.key] = 1
         return output_keys.keys()
+
+    async def _update_progress_and_collect_garbage(
+        self, subtask_graph: SubtaskGraph, interval_seconds: float
+    ):
+        object_ref_to_subtask = self._cur_stage_first_output_object_ref_to_subtask
+        total = len(subtask_graph)
+        completed_subtasks = OrderedSet()
+
+        deletion_seq = []
+
+        def gc():
+            """
+            Consume the completed subtasks and collect garbage.
+
+            GC the output object refs of the subtask which successors are submitted
+            (not completed as above) can reduce the memory peaks, but we can't cancel
+            and rerun slow subtasks because the input object refs of running subtasks
+            may be deleted.
+            """
+            i = 0
+            garbaged_subtasks = set()
+
+            while i < total:
+                while i >= len(completed_subtasks):
+                    yield
+                # Iterate the completed subtasks once.
+                subtask = completed_subtasks[i]
+                i += 1
+                logger.debug("GC: %s", subtask)
+
+                # Note: There may be a scenario in which delayed gc occurs.
+                # When a subtask has more than one predecessor, like A, B,
+                # and in the `for ... in ...` loop we get A firstly while
+                # B's successors are completed, A's not. Then we cannot remove
+                # B's results chunks before A's.
+                for pred in subtask_graph.iter_predecessors(subtask):
+                    while not all(
+                        succ in completed_subtasks
+                        for succ in subtask_graph.iter_successors(pred)
+                    ):
+                        yield
+                    if pred in garbaged_subtasks:
+                        continue
```

Comment:
Checking subtask has been collected should be moved to the beginning of the loop `for pred in subtask_graph.iter_predecessors(subtask)`.

Reply:
Yes, thanks.